### PR TITLE
SIH-SITL integration tests

### PR DESCRIPTION
**Pull request by mbjd**
### Overview

This is an attempt at using SIH-as-SITL for integration tests. All mavsdk integration tests pass, but still not 100% reliably.

This has grown to quite a large collection of changes. Here I summarise the main points. Apologies for the messy nature! If you want help or have any questions, don't hesitate to approach me.

This is mostly a proof of concept. Doubts about the implementation are highlighted with :thinking:. Possibilities to proceed: 
 - cherry pick small, self-contained changes from this to individual PRs
 - re-do this entire PR (or parts of it) from the ground up, learning from this here to arrive at a cleaner implementation
 - merge this PR directly, and improve the implementation subsequently
 - discard it entirely in favour of a deeper overhaul of the integration tests. it would be very desirable to integrate them more tightly with PX4, avoiding the unreliable communication layer in between.

### Entry point 

Tests are launched with the command `test/mavsdk_tests/mavsdk_test_runner.py test/mavsdk_tests/configs/sitl_sih.json`. The file sets `config['simulator'] = sih`, which causes `mavsdk_test_runner.py` to do the rest. Commit: 41c9d1f9d34b77676c4dce7bdd69426fdd7af282

The tests are ran as a github workflow thanks to the [workflow file](https://github.com/PX4/PX4-Autopilot/blob/dd882c29e5683d0d0426751d7b0cfdd9e4e151bc/.github/workflows/sitl_sih_tests.yml). Commit: https://github.com/PX4/PX4-Autopilot/pull/24237/commits/dd882c29e5683d0d0426751d7b0cfdd9e4e151bc

:thinking: hacky implementation with setting env vars in `mavsdk_test_runner.py`
:thinking: bloated yml file with gazebo-related stuff. Not sure if the build system even allows for building tests without gazebo, maybe deeper changes there are needed and it is not worth it to save a couple seconds per CI run. 

### Sensor Failures

Many tests are about sensor failures. For gazebo, these are handled in [`SimulatorMavlink::check_failure_injections()`](https://github.com/PX4/PX4-Autopilot/blob/a9214b3aa3f3e0fdd6b2bd9eb91732a61aa89cdc/src/modules/simulation/simulator_mavlink/SimulatorMavlink.cpp#L1223), with the sensors models being contained in gazebo. However, in SIH, the sensors are simulated separately by modules such as [`SensorGpsSim`](https://github.com/PX4/PX4-Autopilot/blob/a9214b3aa3f3e0fdd6b2bd9eb91732a61aa89cdc/src/modules/simulation/sensor_gps_sim/SensorGpsSim.cpp) etc. Therefore, we implement the sensor failures within those modules. They are mostly a copy of the logic in [`SimulatorMavlink::check_failure_injections()`](https://github.com/PX4/PX4-Autopilot/blob/a9214b3aa3f3e0fdd6b2bd9eb91732a61aa89cdc/src/modules/simulation/simulator_mavlink/SimulatorMavlink.cpp#L1223). They are triggered from the test suite ([example](https://github.com/PX4/PX4-Autopilot/blob/e16bf2203b3e30292e0169fa09bcb8711c79bd40/test/mavsdk_tests/autopilot_tester.cpp#L292C2-L292C7)) but can also be manually triggered from the MAVLink console with the [failure command](https://docs.px4.io/main/en/debug/failure_injection.html).

 Commits: 9a801c6be1df5f869731af98cd062676280ba14e, 4056a5d372a979c0c2dfa99079bab23f2b136af4, 75dd0ceed37b0e1e5387645fddb733db9940ba5e, 55a195cd1fda29f925d0590ba8a428cfccb38dd9, c209d11744ef17c83adbd0ce47520fd880f146cd

:thinking: code repetition... both in terms of repeating logic from `SimulatorMavlink`, but also almost shared logic between the `Sensor...Sim` modules

For the airspeed sensor, there is also an entirely different way to simulate failure, by setting the parameter `SIM_ARSPD_FAIL` (see https://github.com/PX4/PX4-Autopilot/pull/21126). With the implementation of "proper" failures as used in the test suite this is not necessary anymore but is [still possible](https://github.com/PX4/PX4-Autopilot/blob/e16bf2203b3e30292e0169fa09bcb8711c79bd40/src/modules/simulation/sensor_airspeed_sim/SensorAirspeedSim.cpp#L114). 

### Sensor Implementation

SIH itself also implements an [airspeed sensor](https://github.com/PX4/PX4-Autopilot/blob/e16bf2203b3e30292e0169fa09bcb8711c79bd40/src/modules/simulation/simulator_sih/sih.cpp#L559). Here we switch it off (a2a6c35034a1148159bb86c2ed4aee6e0869ebaa) in favour of [`SensorAirspeedSim`](https://github.com/PX4/PX4-Autopilot/blob/a9214b3aa3f3e0fdd6b2bd9eb91732a61aa89cdc/src/modules/simulation/sensor_gps_sim/SensorAirspeedSim.cpp), where we implement the sensor failures. Notice that `SensorAirspeedSim` is only started if the airframe file contains `param set-default SENS_EN_ARSPDSIM 1` (7576725a146, 96105cacc09) and the launch script considers that parameter (55a195cd1fda29f925d0590ba8a428cfccb38dd9). Be sure to only apply those changes together to avoid conflicting or missing sensor data. 

`SensorAirspeedSim` calculated the airspeed previously as dot([1, 0, 0], velocity), corresponding to the assumption that the pitot tube is mounted in body-forward direction. This is incorrect for tailsitters, in which case body-forward points forward in hovering/tail-sitting pose, but _downwards_ in forward flight, where airspeed measurements are needed. As a simple fix, we now instead calculate airspeed as norm(velocity) (same as in SIH). Strictly speaking this is incorrect, as sideways components of the velocity now also contribute to the airspeed, which with an acutal pitot tube measurement would not happen. But in well-behaved fixed wing flight those should be small enough to ignore, and in line with the modelling accuracy standards for SIH generally. Commit: 7576725a146bc1f1c17de9f8a4b8f22f452ec790

### Deterministic Simulation

To exclude the effect of sensor noise, we add a parameter `SIH_NOISE_SCALE` which when set to 0 disables sensor noise. The simulation is still not entirely deterministic due to timing. Commits: 4666a97454d4270b852a80ba533a3aef064d5636, 1470a1cf3bbb986a9b58707f1a1be99b983bfa1e

:thinking: This was mostly implemented as a debugging tool. If merging, we could probably drop it unless someone has a strong need. 
:bulb: Very cool would also be to have a deterministic PRNG for reproducibility. [Jax](https://jax.readthedocs.io/en/latest/random-numbers.html) does this very nicely in the context of multiple threads but with some restrictions (pure functions i.e. no global state). But again, the advantage is probably not worth the effort. 

### Timing bugfixes

When running at large speed factors the tests were previously very unreliable. This was due to timing issues: When [waiting for some condition](https://github.com/PX4/PX4-Autopilot/blob/e16bf2203b3e30292e0169fa09bcb8711c79bd40/test/mavsdk_tests/autopilot_tester.h#L245) or [sleeping for a given amount of simulation time](https://github.com/PX4/PX4-Autopilot/blob/e16bf2203b3e30292e0169fa09bcb8711c79bd40/test/mavsdk_tests/autopilot_tester.h#L168), this is done with real-time polling. However, if running with larger and larger speed factor, polling with the same speed (in real-time) becomes more and more inaccurate. In extreme cases this led to timeouts, with tests failing for example because the VTOL transition took too long, even though the log clearly showed that the transition occured faster than the prescribed timeout. 

We alleviate this by increasing the polling frequency with the speed factor, so it stays equally fast _in simulation time_ across all speed factors (if the simulation is slower than the set speed factor, polling becomes quicker w.r.t. simulation time. This should not be problematic unless we set extremely high, unrealistic speed factors like 10000000, in which case polling will hog the CPU). Commit:  e16bf2203b3e30292e0169fa09bcb8711c79bd40

Also, some tests used real-time sleep statements, leading to obvious timing issues. Adapted to sim-time sleep here: 5a03cfb986e7baefd9837f85e300a252b6f1c857, PR here: https://github.com/PX4/PX4-Autopilot/pull/24266

:thinking: Probably even sim-time sleep statements in tests are not a great idea and there should be better alternatives such as waiting explicitly until the condition we actually depend on is satisfied (or failing if not). 
:thinking: Fundamentally, is polling the right appproach to check conditions like these? Every relevant condition will only change once per simulation step. Can we couple the checks to that? Check not every x (real-time) milliseconds, but once at every lockstep simulation step? Or setup some interrupts for the relevant message directly? 

### Model & parameter changes

It is quite common for airframe files to be self-contradictory because some physical model parameters are redundant (e.g. `SIH_T_MAX` follows from `CA_ROTOR*_CT`, `SIH_Q_MAX` from `CA_ROTOR*_KM`, `SIH_L_ROLL` and `SIH_L_PITCH` from `CA_ROTOR*_PX` and `CA_ROTOR*_PY`). I corrected some of the contradictions here (350f16009053a2c767333d65df0a4c162f33e55f, 96105cacc094487247fbdf94c15c75f6ff3e4b75, 6779c70557fd540c70fb72a8326ffff6b5e91100) but most likely I missed other ones. In the long term it would be very cool to have a minimal, non-redundant set of parameters that avoids this issue altogether and removes an opportunity for the user to mess up. 

Standard VTOL: Magic number in `sih.cpp` is now explained as the factor by which the pusher motor is stronger than the MC motors: a2a6c35034a1148159bb86c2ed4aee6e0869ebaa. Still hardcoded though.

Tailsitter: Channel reversal moved to SIH model to clean up config file: https://github.com/PX4/PX4-Autopilot/commit/dd882c29e5683d0d0426751d7b0cfdd9e4e151bc 

:thinking: this will break all other SIH tailsitter airframes. in the PX4 repo atm there is only 1102, but externally people might have additional ones. If merging, think of some way to handle this smoothly. 

### Test changes

We adapt some test slightly to be more lenient w.r.t. model and tuning variations. 

a2a6c35034a
 - larger position tolerance for MC offboard control
 - Higher takeoff altitude for VTOL Loiter with airspeed failure (standard VTOL)
     - For some configurations, the "wrong" (= too low) airspeed failure causes quite a significant pitch down (caused by TECS wanting to gain airspeed after reading the lower airspeed), before the failure being detected and corrected for. This pitch down is more severe with lower airspeed, and the SIH model flies at lower airspeed than the gazebo one. We also increase airspeed for this airframe in this commit. 
 
:thinking: Passing tests _should_ also be possible to achieve by only changing the model towards the gazebo version which passes tests.... The gazebo models are sometimes *very* different than the SIH ones we are using here in terms of size/mass, so it's not entirely surprising that tests designed for the former sometimes fail with the latter. 

### Stress Test 

Over the weekend I ran the full test suite ~220x to assess reliability, these are the results. The test suite as a whole fails in 2/3 of all runs, but most individual tests are relatively reliable (with some outliers...). I have all the output data including flight logs, if someone wants it please approach me with a thumbdrive/SSD. 

```
Overall test results:
    140  FAIL
     77  PASS

Percentage of failures in total:
230 / 11501 (1.99%)

Number of failures per test:
  - quadx                                                          
     - 'Failure Injection - Reject mid-air when it is disabled'    5
     - 'Land on GPS lost during mission (baro height mode)'        4
     - 'Land on GPS lost during mission (GPS height mode)'         3
     - 'Continue on mag lost during mission'                       7
     - 'Continue on baro lost during mission (baro height mode)'   9
     - 'Continue on baro lost during mission (GPS height mode)'    4
     - 'Continue on baro stuck during mission (baro height mode)'  2
     - 'Continue on baro stuck during mission (GPS height mode)'   6
     - 'Takeoff and Land'                                          2
     - 'Fly square Multicopter Missions including RTL'             6
     - 'Fly square Multicopter Missions with manual RTL'           2
     - 'Fly straight Multicopter Mission'                          34
     - 'Offboard takeoff and land'                                 
     - 'Offboard position control'                                 6
     - 'Offboard attitude control'                                 2
     - 'Fly forward in position control'                           1
     - 'Fly forward in altitude control'                           4
  - standard_vtol                                                  
     - 'Takeoff and Land'                                          4
     - 'Fly forward in position control'                           1
     - 'Fly forward in altitude control'                           
     - 'Fly VTOL mission'                                          5
     - 'Figure eight execution clockwise'                          5
     - 'Figure eight execution counterclockwise'                   6
     - 'RTL direct Home'                                           2
     - 'RTL direct Mission Land'                                   7
     - 'RTL with Mission Landing'                                  5
     - 'RTL with Reverse Mission'                                  2
     - 'RTL direct home without approaches'                        4
     - 'RTL direct home without approaches forced'                 3
     - 'RTL direct home with approaches'                           5
     - 'RTL direct home not as rally point'                        7
     - 'RTL direct rally without approaches'                       1
     - 'RTL direct rally without approaches forced'                5
     - 'RTL direct rally with approaches'                          4
     - 'Fly VTOL mission with wind change (ramped_up_wind.world)'  5
     - 'Fly VTOL Loiter with airspeed failure'                     5
  - xvert                                                          
     - 'Takeoff and Land'                                          3
     - 'Fly forward in position control'                           2
     - 'Fly forward in altitude control'                           
     - 'Fly VTOL mission'                                          5
     - 'Figure eight execution clockwise'                          4
     - 'Figure eight execution counterclockwise'                   5
     - 'RTL direct Home'                                           6
     - 'RTL direct Mission Land'                                   4
     - 'RTL with Mission Landing'                                  2
     - 'RTL with Reverse Mission'                                  
     - 'RTL direct home without approaches'                        5
     - 'RTL direct home without approaches forced'                 6
     - 'RTL direct home with approaches'                           1
     - 'RTL direct home not as rally point'                        3
     - 'RTL direct rally without approaches'                       3
     - 'RTL direct rally without approaches forced'                6
     - 'RTL direct rally with approaches'                          2
```

### CODE REVIEW

**Comment by dakejahl** [comment]
Lots here! We should talk sometime about moving our CI SITL tests to the new gazebo (gz). I'm willing to help get that setup, I know @mrpollo has been looking into getting some new containers setup for gz support. We have a list here https://github.com/PX4/PX4-Autopilot/issues/23602 of missing features but we should figure out what we must have to cover our current suite of tests.
